gitlab/v4/objects/packages.py

```diff
@@ -11,6 +21,110 @@
 ]
 
 
+class GenericPackage(RESTObject):
+    _id_attr = "package_name"
+
+
+class GenericPackageManager(RESTManager):
+    _path = "/projects/%(project_id)s/packages/generic"
+    _obj_cls = GenericPackage
+    _from_parent_attrs = {"project_id": "id"}
+
+    @cli.register_custom_action(
+        "GenericPackageManager",
+        ("package_name", "package_version", "file_name", "path"),
+    )
+    @exc.on_http_error(exc.GitlabUploadError)
+    def upload(
+        self,
+        package_name: str,
+        package_version: str,
+        file_name: str,
+        path: Union[str, Path],
+        **kwargs,
+    ) -> GenericPackage:
+        """Upload a file as a generic package.
+
+        Args:
+            package_name (str): The package name. Must follow generic package
+                                name regex rules
+            package_version (str): The package version. Must follow semantic
+                                version regex rules
+            file_name (str): The name of the file as uploaded in the registry
+            path (str): The path to a local file to upload
+
+        Raises:
+            GitlabConnectionError: If the server cannot be reached
+            GitlabUploadError: If the file upload fails
+            GitlabUploadError: If ``filepath`` cannot be read
+
+        Returns:
+            GenericPackage: An object storing the metadata of the uploaded package.
+        """
+
+        try:
+            with open(path, "rb") as f:
+                file_data = f.read()
+        except OSError:
+            raise exc.GitlabUploadError(f"Failed to read package file {path}")
+
+        url = f"{self._computed_path}/{package_name}/{package_version}/{file_name}"
+        server_data = self.gitlab.http_put(url, post_data=file_data, raw=True, **kwargs)
+
+        return self._obj_cls(
+            self,
+            {
+                "package_name": package_name,
+                "package_version": package_version,
+                "file_name": file_name,
+                "path": path,
+                "message": server_data["message"],
+            },
+        )
+
+    @cli.register_custom_action(
+        "GenericPackageManager",
+        ("package_name", "package_version", "file_name"),
+    )
+    @exc.on_http_error(exc.GitlabGetError)
+    def download(
+        self,
+        package_name: str,
```

Comment:
What would think of:

```suggestion
        *,
        package_name: str,
```

To enforce using keyword arguments.

Reply:
Maybe we should discuss this in a separate issue for all methods as it's not currently enforced in any custom user-facing methods.

Although I feel like this ends up in a lot of ad-hoc/CI scripts and having the convenience of quick one-liners there still has value IMO, e.g. `project.generic_packages.download(package, version, file)` is not too bad?

Reply:
Sounds good. Can discuss later!